File: app/src/main/java/com/meenbeese/chronos/fragments/sound/BaseSoundChooserFragment.java
package com.meenbeese.chronos.fragments.sound;

import android.content.Context;

import androidx.annotation.Nullable;

import java.lang.ref.WeakReference;

import com.meenbeese.chronos.data.SoundData;
import com.meenbeese.chronos.fragments.BasePagerFragment;
import com.meenbeese.chronos.interfaces.ContextFragmentInstantiator;
import com.meenbeese.chronos.interfaces.SoundChooserListener;


public abstract class BaseSoundChooserFragment extends BasePagerFragment implements SoundChooserListener {

    private SoundChooserListener listener;

    public void setListener(SoundChooserListener listener) {
        this.listener = listener;
    }

    @Override
    public void onSoundChosen(SoundData sound) {
        if (listener != null)
            listener.onSoundChosen(sound);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        listener = null;
    }

    abstract static class Instantiator extends ContextFragmentInstantiator {

        private final WeakReference<SoundChooserListener> listener;

        public Instantiator(Context context, SoundChooserListener listener) {
            super(context);
            this.listener = new WeakReference<>(listener);
        }

        @Nullable
        @Override
        public BasePagerFragment newInstance(int position) {
            SoundChooserListener listener = this.listener.get();
            if (listener != null)
                return newInstance(position, listener);
            else return null;
        }

        abstract BasePagerFragment newInstance(int position, SoundChooserListener listener);
    }

}

File: app/src/main/java/com/meenbeese/chronos/fragments/sound/BaseSoundChooserFragment.kt
package com.meenbeese.chronos.fragments.sound

import android.content.Context

import androidx.media3.common.util.UnstableApi

import com.meenbeese.chronos.data.SoundData
import com.meenbeese.chronos.fragments.BasePagerFragment
import com.meenbeese.chronos.interfaces.ContextFragmentInstantiator
import com.meenbeese.chronos.interfaces.SoundChooserListener

@UnstableApi
abstract class BaseSoundChooserFragment : BasePagerFragment(), SoundChooserListener {
    private var listener: SoundChooserListener? = null
    fun setListener(listener: SoundChooserListener?) {
        this.listener = listener
    }

    override fun onSoundChosen(sound: SoundData?) {
        listener?.onSoundChosen(sound)
    }

    override fun onDestroy() {
        super.onDestroy()
        listener = null
    }

    internal abstract class Instantiator(
        context: Context?,
        private val listener: SoundChooserListener?
    ) : ContextFragmentInstantiator(context!!) {

        override fun newInstance(position: Int): BasePagerFragment? {
            return listener?.let { newInstance(position, it) }
        }

        abstract fun newInstance(position: Int, listener: SoundChooserListener?): BasePagerFragment?
    }
}
